### DISCUSSION TITLE
Update benchmarks to include whitelisting checks

### DISCUSSION THREAD

**Discussion by mjethani** [ghostery/adblocker#2170] [Ideas]
From here: https://github.com/cliqz-oss/adblocker/discussions/2165#discussioncomment-1167574

> The question is why neither adblockplus nor ublock apply `genericblock` here?

I have always wondered if the benchmarks were entirely fair, because of all the other things that content blockers do before they finally block or allow a URL request.

Adblock Plus [makes two calls to the engine](https://github.com/adblockplus/adblockpluschrome/blob/1affa87724a7334e589c9a7bb197da8d5e5bf878/lib/requestBlocker.js#L187) _before_ it evaluates the URL itself:

1. Check for a `$document` whitelisting filter for the origin URL
2. Check for a `$genericblock` whitelisting filter for the origin URL

@gorhill it would be nice if you could share the exact steps in uBlock Origin as well.

Now it looks like tsurlfilter already does all of this internally.

For what it's worth, I had proposed that Adblock Plus [do all of these checks in a single call](https://gitlab.com/eyeo/adblockplus/abc/adblockpluscore/-/issues/100), but like so many things it seems to have ended up in a black hole.

Should we update the benchmarks to reflect these additional calls that content blockers must make to check for whitelisting?

**Comment by ameshkov** [comment]
Yeah, I suppose we should do that.

Also, it seems there are already some pretty serious differences that I wasn't aware of (i.e. the absence of `$genericblock` support in ubo), these differences won't go away.

**Comment by remusao** [comment]
So should we use the same mechanism we used to preprocess the lists and remove some filters that are not supported by all blockers? As far as I can see there is only one `$genericblock` filter in the lists used in the benchmark, maybe we can just remove it. Are you aware of any other discrepancy in support for options?

**Reply by ameshkov** [reply]
In the case of a benchmarking tool that would make sense, but these new DEBUG & COMPARE options make it more than just a benchmark. Wouldn't it be better to have an option to observe the differences?

**Reply by remusao** [reply]
Maybe we could have a `COMPAT` mode where we aim at smoothing the differences between engines up to some point, and a normal mode where we just feed the lists as is, depending on what we expect from the results. I would argue that to compare performance it's better to make sure all filters used are supported by all contenders, but debugging would require that we keep everything unfiltered (pun intended) as you suggested in your comment. Wdyt?

**Reply by mjethani** [reply]
Let's leave `DEBUG` and `COMPARE` aside and first ask this question: What are we benchmarking? As far as the matching of requests to filters is concerned, at the moment we are benchmarking a _single_ call to the engine. But is this useful enough on its own? In an actual browser extension, there are multiple calls to the engine _per request_. Shouldn't this be reflected in the benchmarks?

I can tell you from practice that I had my own benchmarks for Adblock Plus and of course they checked for the entire [`webRequest.onBeforeRequest`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onBeforeRequest) handling to the extent that it made sense in Node.js.

**Reply by ameshkov** [reply]
Yep, this is a good point.

**Reply by remusao** [reply]
The intend was initially to capture as much of the entire cost per request as possible. For example in Cliqz the `match(...)` call does everything required to give the final decision for a given request without having to make multiple calls (internally it might query multiple indexes, etc.); the way this is done is considered an implementation detail. If this is not representative of the real life cost for other engines then I guess it would make sense to adapt the benchmark to better reflect the webextension conditions (to the extent this is possible).

I guess what we want to measure is: "What is the cost of deciding to block a request or not?".

If this question is not answered for some of the blockers then maybe we need to update the logic there. Of course there might be overhead induced by other tangential systems like "is the page allowed via the UI by the user", or caching, or keeping track of blocking stats in the extension to display in the UI, so the measurement will never be perfect. But I think focusing on the blocking decision of each engine should be fairly non-ambiguous.

**Reply by remusao** [reply]
On the other hand, it seems like when comparing we get almost 100% the same results between different engines, so I assume what is done on top by other engines are not needed (except for genericblock maybe?) for the lists used in this benchmark.

**Reply by mjethani** [reply]
>I guess what we want to measure is: "What is the cost of deciding to block a request or not?"

Thanks for clarifying the intent.

In that case, I see no issue with adding the whitelisting checks for the content blockers that do such checks outside of the single call to the engine.

> On the other hand, it seems like when comparing we get almost 100% the same results between different engines, so I assume what is done on top by other engines are not needed (except for genericblock maybe?) for the lists used in this benchmark.

For some more context, whitelisting used to be extremely expensive in Adblock Plus until I optimized it (see https://issues.adblockplus.org/ticket/7003/ and https://issues.adblockplus.org/ticket/7089/). For a content blocker that does support such filters, there is a cost. Moreover, among the ones that do support such filters, there could be different ways of implementing it so that one content blocker does better than another. Based on @ameshkov's comments (https://github.com/cliqz-oss/adblocker/discussions/2165#discussioncomment-1167725) it looks like whitelisting filters are a fact of life and not only for business reasons.

> … in Cliqz the `match(...)` call does everything required to give the final decision for a given request

This means that the current benchmarks are not fair to Cliqz and not fair to tsurlfilter either.

**Reply by remusao** [reply]
Sounds good, I guess you are the better person to know how adblockpluscore's integration needs to be adapted. Is it correct to assume that uBlock Origin's measurements already take exceptions into account? (at least that was my understanding).

**Reply by mjethani** [reply]
> Is it correct to assume that uBlock Origin's measurements already take exceptions into account?

The two checks that Adblock Plus does are for `$document` and `$genericblock` and both are [not supported](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#not-supported) by uBlock Origin.

@remusao do Cliqz and Brave support these options?

**Reply by remusao** [reply]
> @remusao do Cliqz and Brave support these options?

No.

From what I can see Brave does not support `$genericblock` and has some support for `$document`. Overall I think the goal was to stick to uBlock Origin's behavior whenever possible.

**Reply by gorhill** [reply]
> The two checks that Adblock Plus does are for `$document` and `$genericblock` and both are [not supported](https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#not-supported) by uBlock Origin.

On the other end uBO does mind filter options not supported by ABP, they may have a cost in uBO even if when a filter list does not use them. For instance, consider the `important` option which requires [an extra step in uBO for _every_ network request](https://github.com/gorhill/uBlock/blob/2aed123b14bf8a8f1cfa82b8bcf97aaa173377de/src/js/static-net-filtering.js#L4196).

In any case, the value of the comparative benchmark here for me is more about whether uBO would seriously diverge in performance from other blockers that I would have to look at what can be improved, and to make the case that webRequest-based blockers can be fast enough that performance is an invalid argument for declarativeNetRequest.

**Reply by mjethani** [reply]
> … to make the case that webRequest-based blockers can be fast enough that performance is an invalid argument for declarativeNetRequest.

Not to go off topic, but I'll just say that Google and eyeo have not been doing this in good faith.

The performance improvements in Adblock Plus were phenomenal. As noted above, whitelisting logic must go over every frame in the hierarchy. This means _several_ additional calls must be made to the matcher. The entire whitelisting logic became incredibly faster as part of the performance work. I did this so Google would not have any excuse. If I remember correctly, request blocking overall became about twenty times faster as part of this work.

Most companies would brag about such remarkable improvements to their software, but eyeo seemed to shove it all under the carpet.

It all adds up.

But these benchmarks can still be used to convince others who may still be making arguments in good faith.

**Reply by mjethani** [reply]
>But these benchmarks can still be used to convince others who may still be making arguments in good faith.

To my knowledge, Google has never presented any data about performance.

But the proof is in the pudding. I stand ready to benchmark any `declarativeNetRequest`-based content blocker on Chrome with uBlock Origin on Firefox on the same system. We can look at the overall CPU and memory usage. We can look at various other measures. I'd be willing to bet that uBlock Origin will win hands down.

**Comment by mjethani** [comment]
Unfortunately the way the whitelisting logic works is that we need to check not only the frame URL but also the URLs of all of the ancestor frames. This means we would need the entire frame hierarchy in the data set. We do not have this at the moment.

@ameshkov why does tsurlfilter only care about the current frame? Is there a way to disable the `$document` and `$genericblock` checks in tsurlfilter for now?

We could of course remove all such filters.

**Reply by remusao** [reply]
Out of curiosity, why do you need intermediary frames, isn't it the intent usually to check only the frame URL and/or the URL of the tab? I guess it would get things wrong if there are iframes embedded into each other or?

**Reply by mjethani** [reply]
The parent can be a [`data:` or `about:blank` frame](https://github.com/adblockplus/adblockpluschrome/blob/1affa87724a7334e589c9a7bb197da8d5e5bf878/lib/url.js#L24).

Also the filter `@@||example.com^$document` should whitelist any `example.com` document, _including any subdocuments_. If `example.com` loads any frames from `example.net` and those frames make requests that might otherwise be blocked, they should be whitelisted too.

**Reply by remusao** [reply]
Ok, my insights are not the freshest at this point but I would assume that to allow a request from `example.com` loaded as an iframe the type should be `$subdocument` or `$frame` and not `$document`.

**Reply by mjethani** [reply]
In the above example `example.com` is the top frame whereas `example.net` is a subframe. The filter `@@||example.com^$document` should still whitelist requests from `example.net` — if it's a subframe of `example.com`. The target of the filter here is `example.com`, which is the top frame.

**Reply by mjethani** [reply]
>@ameshkov why does tsurlfilter only care about the current frame?

Alright, I take it that tsurlfilter assumes that the current frame is the top frame. In this case, we can also assume the same for Adblock Plus and do a one-level check.

This does not affect any of the other content blockers based on the comments above.

Note that adding the whitelisting logic will make the results for Adblock Plus similar to those for tsurlfilter and then they'll be different from uBlock Origin, Cliqz, and Brave.

We could still remove the `$document` and `$genericblock` filters but that would be orthogonal to the whitelisting logic.

**Reply by ameshkov** [reply]
> @ameshkov why does tsurlfilter only care about the current frame? Is there a way to disable the $document and $genericblock checks in tsurlfilter for now?

Please don't do that just yet. The API is not final and will probably change by the time we integrate it into the extension. I'll see to us taking care of this problem as well.

**Comment by mjethani** [comment]
This is now approved and merged here: https://github.com/cliqz-oss/adblocker/pull/2172
